api/projects/code_references/services.py

```diff
@@ -0,0 +1,52 @@
+from datetime import timedelta
+
+from django.contrib.postgres.expressions import ArraySubquery
+from django.db.models import F, Func, OuterRef, QuerySet, Value
+from django.db.models.functions import JSONObject
+from django.utils import timezone
+
+from features.models import Feature
+from projects.code_references.models import FeatureFlagCodeReferencesScan
+
+
+def annotate_feature_queryset_with_code_references_summary(
+    queryset: QuerySet[Feature],
+) -> QuerySet[Feature]:
+    """Extend feature objects with a `code_references_counts`
+
+    NOTE: This adds compatibility with `CodeReferenceRepositoryCountSerializer`
+    while preventing N+1 queries from the serializer.
+    """
+    counts_by_repository = (
+        FeatureFlagCodeReferencesScan.objects
+        # Count code references from JSON matching the feature name
+        .annotate(
+            feature_name=OuterRef("name"),
+            count=Func(
+                Func(
+                    F("code_references"),
+                    Value("$[*] ? (@.feature_name == $feature_name)"),
```

Comment:
Minor: I may be wrong, but GIN expression indexes usually require the query to match exactly the index (`$[*].feature_name`) from what I recall. Mind running EXPLAIN ANALYZE to verify?